### ISSUE TITLE
Is too-few-public-methods (R0903 ) really useful?

### ISSUE THREAD

**Issue by uriva** [PyCQA/pylint#2710]
This message over triggers many times, e.g. for dataclasses (or at least the backport to python 3.6).

See other examples in https://stackoverflow.com/questions/14027417/what-does-pylints-too-few-public-methods-message-mean

Consider removing altogether.

**Comment by matkoniecz** [comment]
Can you provide examples here? External sites may be changed so link only reports are not too useful.

**Comment by matkoniecz** [comment]
And first example that I spotted, in https://stackoverflow.com/a/14027686/4130619 is a case where complain makes a perfect sense.

**Comment by PCManticore** [comment]
@uriva As mentioned by @matkoniecz in most cases the check makes sense. If your class has a single method on it, it's probably a function in disguise. You can also configure the number of public methods with [min-public-methods](http://pylint.pycqa.org/en/latest/technical_reference/features.html#design-checker-options). But there are a couple of classes for which we should not report this, as it makes perfect sense to not have too many methods or no methods at all. This includes `attrs`, `dataclasses`, `typing.NamedTuple` and similar. 
Given that your issue is about the removal of the method, I'm going to close this issue, but if you have an example where `pylint` complains about a dataclass, please open a separate issue.

**Comment by uriva** [comment]
Ok.
I think it would make sense to send back warnings people disable to make this discussion more objective.

**Comment by g1itch** [comment]
```patch
diff --git a/src/messagetypes/__init__.py b/src/messagetypes/__init__.py
index 06783eac..436eef14 100644
--- a/src/messagetypes/__init__.py
+++ b/src/messagetypes/__init__.py
@@ -1,14 +1,25 @@
+import os
+from abc import ABCMeta, abstractmethod
 from importlib import import_module
-from os import path, listdir
-from string import lower
 
-from debug import logger
 import messagetypes
 import paths
+from debug import logger
+
 
 class MsgBase(object):
-    def encode(self):
-        self.data = {"": lower(type(self).__name__)}
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self.subject = self.body = self.data = None
+
+    @abstractmethod
+    def encode(self, data):
+        self.data = {"": type(self).__name__.lower()}
+
+    @abstractmethod
+    def decode(self, data):
+        pass
 
 
 def constructObject(data):
@@ -18,7 +29,9 @@ def constructObject(data):
     try:
         classBase = getattr(getattr(messagetypes, data[""]), data[""].title())
     except (NameError, AttributeError):
-        logger.error("Don't know how to handle message type: \"%s\"", data[""], exc_info=True)
+        logger.error(
+            "Don't know how to handle message type: \"%s\"",
+            data[""], exc_info=True)
         return None
     try:
         returnObj = classBase()
@@ -32,18 +45,19 @@ def constructObject(data):
     else:
         return returnObj
 
-if paths.frozen is not None:
+
+if paths.frozen:
     import messagetypes.message
     import messagetypes.vote
 else:
-    for mod in listdir(path.dirname(__file__)):
+    for mod in os.listdir(os.path.dirname(__file__)):
         if mod == "__init__.py":
             continue
-        splitted = path.splitext(mod)
-        if splitted[1] != ".py":
+        name, ext = os.path.splitext(mod)
+        if ext != ".py":
             continue
         try:
-            import_module(".{}".format(splitted[0]), "messagetypes")
+            import_module(".{}".format(name), "messagetypes")
         except ImportError:
             logger.error("Error importing %s", mod, exc_info=True)
         else:
diff --git a/src/messagetypes/message.py b/src/messagetypes/message.py
index f52c6b35..dda82955 100644
--- a/src/messagetypes/message.py
+++ b/src/messagetypes/message.py
@@ -3,22 +3,19 @@ from messagetypes import MsgBase
 
 
 class Message(MsgBase):
-    def __init__(self):
-        return
-
     def decode(self, data):
         # UTF-8 and variable type validator
-        if type(data["subject"]) is str:
+        if isinstance(data["subject"], str):
             self.subject = unicode(data["subject"], 'utf-8', 'replace')
         else:
             self.subject = unicode(str(data["subject"]), 'utf-8', 'replace')
-        if type(data["body"]) is str:
+        if isinstance(data["body"], str):
             self.body = unicode(data["body"], 'utf-8', 'replace')
         else:
             self.body = unicode(str(data["body"]), 'utf-8', 'replace')
 
     def encode(self, data):
-        super(Message, self).encode()
+        super(Message, self).encode(data)
         try:
             self.data["subject"] = data["subject"]
             self.data["body"] = data["body"]
```

**Comment by kuwv** [comment]
@PCManticore there is one false positive where this really shouldn't.

Descriptors contain only dunders (which are not technically private methods anyway since many can be called directly):

```
class Example:
    __get__(self, obj: object, owner: Optional[Type[object]] = None) -> T:
        ...
    __set__(self, obj: object, value: T) -> None:
        ...
```

I don't remember is single-method classes trigger but a lot of common OOP design patterns use these. (Additionally mixins can be lumped in this as well) This is probably fine to trigger on though but IMO a bit inaccurate in Python design scope of OOP first language.


```
class Converter(AbstractImpl):
    def convert(self, content: Dict[str, Any]) -> str:
        ...
```

**Comment by matkoniecz** [comment]
it may be better to open new issue over commenting in 2019 one

**Comment by kuwv** [comment]
Ok, I'll create a new issue and reference this for @uriva.
